File: android/app/src/test/java/com/vpn/android/p2p/SignalEnvelopeTest.java
package com.vpn.android.p2p;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

/**
 * The signal envelope (docs/research/P2P_RELAY_FEASIBILITY.md §8.1) is a
 * wire contract shared with independently-built Desktop/Linux relay agents —
 * these tests pin the exact JSON shape so a change here can't silently drift
 * from what those other implementations expect.
 */
public class SignalEnvelopeTest {

    @Test
    public void offerRoundTrips() {
        SignalEnvelope original = SignalEnvelope.offer("v=0...", "203.0.113.7", 443);
        SignalEnvelope parsed = SignalEnvelope.parse(original.toBytes());

        assertEquals(SignalEnvelope.KIND_OFFER, parsed.kind);
        assertEquals("v=0...", parsed.sdp);
        assertEquals("203.0.113.7", parsed.targetHost);
        assertEquals(443, parsed.targetPort);
    }

    @Test
    public void answerRoundTrips() {
        SignalEnvelope original = SignalEnvelope.answer("v=0 answer...");
        SignalEnvelope parsed = SignalEnvelope.parse(original.toBytes());

        assertEquals(SignalEnvelope.KIND_ANSWER, parsed.kind);
        assertEquals("v=0 answer...", parsed.sdp);
    }

    @Test
    public void iceCandidateRoundTrips() {
        SignalEnvelope original = SignalEnvelope.ice("candidate:1 1 UDP ...", "0");
        SignalEnvelope parsed = SignalEnvelope.parse(original.toBytes());

        assertEquals(SignalEnvelope.KIND_ICE, parsed.kind);
        assertEquals("candidate:1 1 UDP ...", parsed.candidate);
        assertEquals("0", parsed.sdpMid);
    }

    @Test
    public void parseReturnsNull_forGarbageBytes() {
        assertNull(SignalEnvelope.parse("not json at all".getBytes()));
    }

    @Test
    public void parseReturnsNull_forJsonMissingKind() {
        assertNull(SignalEnvelope.parse("{\"sdp\":\"v=0\"}".getBytes()));
    }

    @Test
    public void exactWireFormat_offer() {
        // Pinned literal shape — any independently-built peer (Desktop/Linux)
        // must produce/accept exactly this, field names included.
        String json = new String(SignalEnvelope.offer("SDP", "1.2.3.4", 9000).toBytes());
        assertEquals("{\"kind\":\"offer\",\"sdp\":\"SDP\",\"targetHost\":\"1.2.3.4\",\"targetPort\":9000}", json);
    }
}

File: desktop/test/signalEnvelope.test.ts
import { describe, expect, it } from 'vitest';
import { decodeSignal, encodeSignal, type SignalEnvelope } from '../src/main/p2p/signalEnvelope';

// This exact wire format (docs/research/P2P_RELAY_FEASIBILITY.md §8.1) is
// shared verbatim across the Desktop/Android/Linux-agent relay
// implementations, built independently in parallel — a round-trip test here
// is the only thing standing between "interoperates" and "silently doesn't".
describe('signalEnvelope', () => {
  it.each<SignalEnvelope>([
    { kind: 'offer', sdp: 'v=0\r\no=- 1 IN IP4 0.0.0.0', targetHost: '203.0.113.7', targetPort: 443 },
    { kind: 'answer', sdp: 'v=0\r\no=- 2 IN IP4 0.0.0.0' },
    { kind: 'ice', candidate: 'candidate:1 1 UDP 12345 1.2.3.4 5000 typ host', sdpMid: '0' },
  ])('round-trips %o', (envelope) => {
    expect(decodeSignal(encodeSignal(envelope))).toEqual(envelope);
  });

  it('rejects a payload missing "kind"', () => {
    expect(() => decodeSignal(Buffer.from('{}'))).toThrow();
  });

  it('rejects an unknown "kind"', () => {
    expect(() => decodeSignal(Buffer.from(JSON.stringify({ kind: 'bogus' })))).toThrow();
  });

  it('rejects an "offer" missing targetPort', () => {
    expect(() => decodeSignal(Buffer.from(JSON.stringify({ kind: 'offer', sdp: 'x', targetHost: 'h' })))).toThrow();
  });

  it('rejects malformed JSON entirely', () => {
    expect(() => decodeSignal(Buffer.from('not json'))).toThrow();
  });

  it('accepts a Uint8Array the same as a Buffer', () => {
    const envelope: SignalEnvelope = { kind: 'answer', sdp: 'x' };
    const encoded = new Uint8Array(encodeSignal(envelope));
    expect(decodeSignal(encoded)).toEqual(envelope);
  });
});
